### ISSUE TITLE
mat1 and mat2 must have the same dtype, but got BFloat16 and Float

### ISSUE THREAD

**Issue by hjj-lmx** [facebookresearch/sam2#577]
# 初始化推理状态
        inference_state = sam2_model.video_predictor.init_state(video_path=input_video_path)

        # 传入的是多帧的mask数组
        for mask_object in mask_objects:
            mask_byte = down_file.url_to_byte(mask_object.get("mask_url"))
            image_data = io.BytesIO(mask_byte)
            mask_image = Image.open(image_data).convert('L')  # 转换为灰度图
            transform = transforms.Compose([
                transforms.ToTensor(),
                transforms.Lambda(lambda x: (x > 0.5).float())
            ])
            mask = transform(mask_image)
            mask = mask.squeeze(0)
            sam2_model.video_predictor.add_new_mask(inference_state,
                                                    mask_object.get("frame_idx"),
                                                    mask_object.get("frame_idx"),
                                                    mask)
        # 传播提示获取整个视频的masklet
        video_segments = {}
        for out_frame_idx, out_obj_ids, out_mask_logits in sam2_model.video_predictor.propagate_in_video(
                inference_state):
            video_segments[out_frame_idx] = {
                out_obj_id: (out_mask_logits[i] > 0.0).cpu().numpy()
                for i, out_obj_id in enumerate(out_obj_ids)
            }
mask_objects的长度是1的适合能正常运行，大于1就报错

**Comment by kascesar** [comment]
Same here!

**Comment by kascesar** [comment]
i had solve the problem with "cpu". Add the nex line:

torch.autocast("cpu", dtype=torch.bfloat16).__enter__()

**Comment by mad4octos** [comment]
> i had solve the problem with "cpu". Add the nex line:
> 
> torch.autocast("cpu", dtype=torch.bfloat16).**enter**()

I'm having the same issue here. Where did you add this line of code to resolve this error? Thanks!! @kascesar

**Comment by kascesar** [comment]
`if torch.cuda.is_available():
    device = torch.device("cuda")
elif torch.backends.mps.is_available():
    device = torch.device("mps")
else:
    device = torch.device("cpu")
print(f"using device: {device}")

if device.type == "cuda":
    # use bfloat16 for the entire notebook
    torch.autocast("cuda", dtype=torch.bfloat16).__enter__()
    # turn on tfloat32 for Ampere GPUs (https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices)
    if torch.cuda.get_device_properties(0).major >= 8:
        torch.backends.cuda.matmul.allow_tf32 = True
        torch.backends.cudnn.allow_tf32 = True

elif device.type == "mps":
    print(
        "\nSupport for MPS devices is preliminary. SAM 2 is trained with CUDA and might "
        "give numerically different outputs and sometimes degraded performance on MPS. "
        "See e.g. https://github.com/pytorch/pytorch/issues/84936 for a discussion."
    )

elif device.type == "cpu":
    torch.autocast("cpu", dtype=torch.bfloat16).__enter__()`

This Is what i alredy use After import bro

**Comment by mad4octos** [comment]
Interesting, for me that produces an "AttributeError 'autocast' object has no attribute 'enter'" and doesn't resolve the issue. You don't have this problem @kascesar ?

**Comment by heyoeyo** [comment]
> "AttributeError 'autocast' object has no attribute 'enter'"

It's probably meant to be `__enter__()` but the double underscore is interpreted as making things __bold__ in markdown.

More generally, this sort of error occurs if the model is loaded using one data type (e.g. bfloat16) whereas the input data is using a different data type (e.g. float). You can fix the problem by making sure the data is loaded on the same device/dtype as the model (or use autocast, which has a similar effect). The error message would usually give some indication of where this is happening. In the original post, the line:
```python
transforms.Lambda(lambda x: (x > 0.5).float())
```
Is an example that might be causing issues, since the data is being cast with `.float()`, but the model is probably using bfloat16. One solution would be to cast it using `.to(device, dtype)`, for example:
```python
transforms.Lambda(lambda x: (x > 0.5).to("cuda", torch.bfloat16))
```
(assuming the model is loaded on a cuda device using bfloat16)

**Comment by kascesar** [comment]
Hi bro! @mad4octos The github parsing convert to bold`__enter__`, just change __enter__ by `__enter__`
